cpmpy/expressions/globalfunctions.py

```diff
@@ -285,8 +287,19 @@ def decompose_comparison(self, cpm_op, cpm_rhs):
 
         """
         arr, idx = self.args
-        return [(idx == i).implies(eval_comparison(cpm_op, arr[i], cpm_rhs)) for i in range(len(arr))] + \
-               [idx >= 0, idx < len(arr)], []
+        cons = []
+        # the decomposition posts `idx=i -> arr[i] <CMP_OP> cpm_rhs`
+        # for every  `i` in the intersection of array indices and the bounds of `idx`
+        lb, ub = idx.get_bounds()
```

Comment:
This will go wrong when idx is not an expression (see prev comment). Better to use `utils.get_bounds(idx)` instead.

Reply:
Ah, definitely the function I was looking for! Thanks